pkg/store/initializers.go

```diff
@@ -0,0 +1,125 @@
+package store
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	grpc "github.com/go-skynet/LocalAI/pkg/grpc"
+	"github.com/phayes/freeport"
+	"github.com/rs/zerolog/log"
+)
+
+var Aliases map[string]string = map[string]string{
+	"embedded": "local",
+}
+
+const (
+	LocalEmbedded = "local"
+)
+
+// starts the grpcStoreProcess for the backend, and returns a grpc client
+// It also loads the store
+func (ml *StoreLoader) grpcStore(backend string, o *Options) func(string) (StoreAddress, error) {
+	return func(storeName string) (StoreAddress, error) {
+		log.Debug().Msgf("Loading Store %s with gRPC (backend: %s): %+v", storeName, backend, *o)
+
+		var client StoreAddress
+
+		getFreeAddress := func() (string, error) {
+			port, err := freeport.GetFreePort()
+			if err != nil {
+				return "", fmt.Errorf("failed allocating free ports: %s", err.Error())
+			}
+			return fmt.Sprintf("127.0.0.1:%d", port), nil
+		}
+
+		grpcProcess := filepath.Join(o.assetDir, "backend-assets", "grpc", "stores", backend)
+		// Check if the file exists
+		if _, err := os.Stat(grpcProcess); os.IsNotExist(err) {
+			return "", fmt.Errorf("grpc process not found: %s. some backends(stablediffusion, tts) require LocalAI compiled with GO_TAGS", grpcProcess)
+		}
+
+		serverAddress, err := getFreeAddress()
+		if err != nil {
+			return "", fmt.Errorf("failed allocating free ports: %s", err.Error())
+		}
+
+		// Make sure the process is executable
+		if err := ml.startProcess(grpcProcess, o.store, serverAddress); err != nil {
+			return "", err
+		}
+
+		log.Debug().Msgf("GRPC Service Started")
+
+		client = StoreAddress(serverAddress)
+
+		// Wait for the service to start up
+		ready := false
+		for i := 0; i < o.grpcAttempts; i++ {
+			alive, err := client.GRPC(o.parallelRequests).HealthCheck(context.Background())
+			if alive {
+				log.Debug().Msgf("GRPC Service Ready")
+				ready = true
+				break
+			}
+			if err != nil && i == o.grpcAttempts-1 {
+				log.Error().Msgf("Failed starting/connecting to the gRPC service: %s", err.Error())
+			}
+			time.Sleep(time.Duration(o.grpcAttemptsDelay) * time.Second)
+		}
+
+		if !ready {
+			log.Debug().Msgf("GRPC Service NOT ready")
+			return "", fmt.Errorf("grpc service not ready")
+		}
+
+		// TODO: This could be where we load storeName if it is persistent
+
+		return client, nil
+	}
+}
+
+func (ml *StoreLoader) resolveAddress(addr StoreAddress, parallel bool) (grpc.Backend, error) {
+	if parallel {
+		return addr.GRPC(parallel), nil
+	}
+
+	if _, ok := ml.grpcClients[string(addr)]; !ok {
+		ml.grpcClients[string(addr)] = addr.GRPC(parallel)
+	}
+	return ml.grpcClients[string(addr)], nil
+}
+
+func (ml *StoreLoader) BackendLoader(opts ...Option) (client grpc.Backend, err error) {
```

Comment:
even if it might be easier duplicating here - it would be fair imho to try generalizing the current ModelLoader. 
There is quite some overlap here, right? or is there something relatively big in terms of changes?

if it's just because it is of the naming, I'd suggest to just refactor it out and call it "backend loader"

Reply:
In my original thinking a store or database doesn't have a model and a model doesn't have the concept of getting or setting exact data. However there are definitely models and databases that blur the line.

Indeed these models with 1-10M token context lengths, I do wonder if that is due to some clever indexing and retrieval.

So yeah I guess generalizing and renaming the model loader is the way to go.

I wonder though if the backend loader should be sub divided somehow. There is the code interpreter and other types of tool that assistants can use. These could be other backends in addition to retrieval.